### PR TITLE
Feature/mx 1502 wikdiata results refactor

### PR DESCRIPTION
**Pull request by mr-kamran-ali**
# Changes
- return only one organization from wikidata. If multiple or no organization is found then return None
- filter quotation marks (") from requested wikidata label (MX-1526)

### CODE REVIEW

File: mex/common/wikidata/extract.py [hunk 1]
```diff
@@ -22,6 +20,7 @@ def search_organization_by_label(
         Generator for WikidataOrganization items
```
Comment (cutoffthetop) [inline mex/common/wikidata/extract.py:20]: needs update

File: mex/common/wikidata/extract.py [hunk 2]
```diff
@@ -54,7 +54,12 @@ def search_organization_by_label(
                 f"IndexError: Error processing results for {item_label}"
             ) from exc
```
Comment (cutoffthetop) [inline mex/common/wikidata/extract.py:55]: why are we catching `requests.exceptions` errors **here** and not around `_get_organization_details`?

coverage report already gives a good hint for that issue:

```
Name                                     Stmts   Miss Branch BrPart  Cover   Missing
------------------------------------------------------------------------------------
mex/common/wikidata/extract.py              29      4     10      3    82%   41, 45, 52-53
```

File: mex/common/wikidata/extract.py [hunk 3]
```diff
@@ -54,7 +54,12 @@ def search_organization_by_label(
...
 
-        yield _get_organization_details(wd_item_id)
+        resolved_organizations.append(_get_organization_details(wd_item_id))
+
+    if len(resolved_organizations) == 1:
```
Comment (cutoffthetop) [inline mex/common/wikidata/extract.py:59]: not very efficient. we could bail out with `None` as soon as we find a second org in `results` - no need to finish the loop

File: tests/wikidata/test_extract.py [hunk 4]
```diff
@@ -18,12 +18,20 @@
 @pytest.mark.integration
 def test_search_organization_by_label() -> None:
     """Test organization search in wikidata."""
-    expected = "Q26678"
+    expected = "Q679041"
 
-    search_results = list(search_organization_by_label(item_label="BMW"))
+    search_results = search_organization_by_label(item_label='Robert Koch Institute"')
```
Comment (cutoffthetop) [inline tests/wikidata/test_extract.py:23]: better example 👍🙂 

but the variable should now be named `search_result`, its just one org (or None)

File: tests/wikidata/test_extract.py [hunk 5]
```diff
@@ -18,12 +18,20 @@
...
+    if search_results:
+        assert search_results.identifier == expected
+    else:
+        pytest.fail(f"No organizations were found for id: {expected}")
```
Comment (cutoffthetop) [inline tests/wikidata/test_extract.py:27]: why pytest.fail, and not simply `assert search_result`?

File: tests/wikidata/test_extract.py [hunk 6]
```diff
@@ -18,12 +18,20 @@
...
 
-    assert len(search_results) == 2
-    assert search_results[0].identifier == expected
+
+@pytest.mark.integration
+def test_search_organization_by_label_for_none() -> None:
+    """Test organization search in wikidata."""
```
Comment (cutoffthetop) [inline tests/wikidata/test_extract.py:32]:
```suggestion
```

no need for docstrings in tests, if they don't tell us more than the test name itself (here it has even less informational content)

File: tests/wikidata/test_extract.py [hunk 7]
```diff
@@ -156,6 +164,47 @@ def test_search_organization_by_label_mocked(monkeypatch: MonkeyPatch) -> None:
...
+        "get_wikidata_item_details_by_id",
+        lambda self, _: mocked_item_details_response(),
+    )
+
+    search_results = search_organization_by_label(item_label="TEST")
+
+    if search_results:
+        assert search_results.model_dump() == expected_item_details_response
+    else:
+        pytest.fail(
```
Comment (cutoffthetop) [inline tests/wikidata/test_extract.py:198]: see above

File: tests/wikidata/test_extract.py [hunk 8]
```diff
@@ -185,10 +234,9 @@ def mocked_item_details_response():
         lambda self, _: mocked_item_details_response(),
     )
 
-    search_results = list(search_organization_by_label(item_label="TEST"))
+    search_results = search_organization_by_label(item_label="TEST")
```
Comment (cutoffthetop) [inline tests/wikidata/test_extract.py:237]:
```suggestion
    search_result = search_organization_by_label(item_label="TEST")
```

File: tests/wikidata/test_transform.py [hunk 9]
```diff
@@ -62,9 +64,11 @@ def test_transform_wikidata_organization_to_organization(
         )
     )
 
-    assert len(extracted_organizations) == 1
+    if not extracted_organizations:
+        pytest.fail("No extracted organization.")
```
Comment (cutoffthetop) [inline tests/wikidata/test_transform.py:68]:
```suggestion
    assert extracted_organizations
```

**Review by cutoffthetop — CHANGES_REQUESTED** [review]
so far, so neat. some fine-tuning needed though.

and: since this is a breaking change, we should prepare a release after this is merged

File: mex/common/wikidata/extract.py (hunk 3 above)
Comment (mr-kamran-ali) [reply mex/common/wikidata/extract.py]: Good point, fixed it in 443c1a2a28dea9bc5394d5e2eeff3320543c0e02

File: tests/wikidata/test_transform.py (hunk 9 above)
Comment (mr-kamran-ali) [reply tests/wikidata/test_transform.py]: updated in 108989566acf52f31953594e8d8cbd61be22be7a

File: tests/wikidata/test_extract.py (hunk 8 above)
Comment (mr-kamran-ali) [reply tests/wikidata/test_extract.py]: updated in 108989566acf52f31953594e8d8cbd61be22be7a

File: tests/wikidata/test_extract.py (hunk 7 above)
Comment (mr-kamran-ali) [reply tests/wikidata/test_extract.py]: updated in 108989566acf52f31953594e8d8cbd61be22be7a

File: tests/wikidata/test_extract.py (hunk 6 above)
Comment (mr-kamran-ali) [reply tests/wikidata/test_extract.py]: updated in 108989566acf52f31953594e8d8cbd61be22be7a

File: tests/wikidata/test_extract.py (hunk 5 above)
Comment (mr-kamran-ali) [reply tests/wikidata/test_extract.py]: I was trying to be very explicit. 😩

File: tests/wikidata/test_extract.py (hunk 4 above)
Comment (mr-kamran-ali) [reply tests/wikidata/test_extract.py]: updated in 108989566acf52f31953594e8d8cbd61be22be7a

**Comment by mr-kamran-ali** [comment]
> and: since this is a breaking change, we should prepare a release after this is merged

I was planning to do that after this merge. then this new version will be used in mex-extractors

File: mex/common/wikidata/extract.py (hunk 2 above)
Comment (cutoffthetop) [reply mex/common/wikidata/extract.py]: but **how** would `wd_item_id = results[0]["item"]["value"].split("/")[-1]` produce an `HTTPError`?

File: mex/common/wikidata/extract.py (hunk 3 above)
Comment (cutoffthetop) [reply mex/common/wikidata/extract.py]: this commit changes the interface of this function though. before, if you had 0 results, you'd still get a `None`, now you get a `MExError`

File: tests/wikidata/test_extract.py [hunk 10]
```diff
@@ -156,6 +158,44 @@ def test_search_organization_by_label_mocked(monkeypatch: MonkeyPatch) -> None:
...
+        "get_wikidata_item_details_by_id",
+        lambda self, _: mocked_item_details_response(),
+    )
+
+    search_result = search_organization_by_label(item_label="TEST")
+
+    if search_result:
+        assert search_result.model_dump() == expected_item_details_response
+
+    assert search_result
```
Comment (cutoffthetop) [inline tests/wikidata/test_extract.py:192]:
```suggestion
    assert search_result
    assert search_result.model_dump() == expected_item_details_response
```

File: tests/wikidata/test_extract.py (hunk 5 above)
Comment (cutoffthetop) [reply tests/wikidata/test_extract.py]: ok, that is always good of course ;) just the `pytest.fail` syntax is not very intuitive. if you'd like to provide a custom error message to an assertion you can still do that with the `assert` syntax. just add a `,` and your message:

```python
assert search_results, f"No organizations were found for id: {expected}"
assert search_results.identifier == expected
```

File: mex/common/wikidata/extract.py (hunk 3 above)
Comment (mr-kamran-ali) [reply mex/common/wikidata/extract.py]: What do you think about a simpler solution like this: c555328cf8ee1896705dcbebab1727453fb9b872

File: tests/wikidata/test_extract.py (hunk 5 above)
Comment (mr-kamran-ali) [reply tests/wikidata/test_extract.py]: this is neater. I like it. implemented.

File: mex/common/wikidata/extract.py (hunk 2 above)
Comment (mr-kamran-ali) [reply mex/common/wikidata/extract.py]: thats a good question which I dont know the answer of. Maybe when we were working with generators the call was made when it tried to access something in the `results` list. But now its irrelevant, and I removed it.

File: mex/common/wikidata/extract.py (hunk 3 above)
Comment (cutoffthetop) [reply mex/common/wikidata/extract.py]: yup, that works as well.

but: the IndexError can not happen anymore, so you could trash that. it was not covered by tests anyway

```
 Name                                     Stmts   Miss Branch BrPart  Cover   Missing
------------------------------------------------------------------------------------
mex/common/wikidata/extract.py              21      2      4      1    88%   42-43
```

**Review by cutoffthetop — CHANGES_REQUESTED** [review]
alright, just a few [lines to delete](https://github.com/robert-koch-institut/mex-common/pull/127/files/7657887a31ca78d701aaf84bcb02d9a2ffb05f46#r1555909882) and a merge of the main branch, then we're ready to go here :)

**Review by cutoffthetop — APPROVED** [review]
🚢

**Pull request merged**